pkg/agent/cniserver/pod_monitor.go

```diff
@@ -0,0 +1,246 @@
+	// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+	if status.Desc.State == openflow15.PS_LIVE {
+		ovsPort := string(bytes.Trim(status.Desc.Name, "\x00"))
+		ofPort := status.Desc.PortNo
+		m.statusMapMtx.Lock()
+		m.ovsPortStatusMap[ovsPort] = ofPort
+		m.statusMapMtx.Unlock()
+		m.queue.Add(ovsPort)
+	}
+}
+
+func (m *podIfaceMonitor) processNextWorkItem() bool {
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	defer m.queue.Done(key)
+
+	if err := m.updateUnreadyPod(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		m.queue.Forget(key)
+
+		m.statusMapMtx.Lock()
+		defer m.statusMapMtx.Unlock()
+		delete(m.ovsPortStatusMap, key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		m.queue.AddAfter(key, retryInterval)
+	}
+	return true
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the workqueue.
+func (m *podIfaceMonitor) worker() {
+	for m.processNextWorkItem() {
+	}
+}
+
+func (m *podIfaceMonitor) updatePodFlows(ifName string, ofPort int32, podInfo *unreadyPodInfo) error {
+	ifConfig, found := m.ifaceStore.GetInterfaceByName(ifName)
+	if !found {
+		klog.InfoS("Interface config is not found in local cache, remove from local cache", "port", ifName)
+		delete(m.unreadyInterfaces, ifName)
+		return nil
+	}
+
+	// Update interface config with the ofPort.
+	ifConfig.OVSPortConfig.OFPort = ofPort
+	m.ifaceStore.UpdateInterface(ifConfig)
+
+	// Install OpenFlow entries for the Pod.
+	klog.V(2).InfoS("Setting up Openflow entries for OVS port", "port", ifName)
+	if err := m.ofClient.InstallPodFlows(ifName, ifConfig.IPs, ifConfig.MAC, uint32(ofPort), ifConfig.VLANID, nil); err != nil {
+		return fmt.Errorf("failed to add Openflow entries for OVS port %s: %v", ifName, err)
+	}
+
+	// Notify the Pod update event to required components.
+	event := agenttypes.PodUpdate{
+		PodName:      ifConfig.PodName,
+		PodNamespace: ifConfig.PodNamespace,
+		IsAdd:        true,
+		ContainerID:  ifConfig.ContainerID,
+	}
+	m.podUpdateNotifier.Notify(event)
+
+	podInfo.flowsInstalled = true
+	// Update Pod event immediately
+	m.processPodEvents(ifName, podInfo)
+	return nil
+}
+
+func (m *podIfaceMonitor) updateUnreadyPod(ovsPort string) error {
+	m.unreadyMapMutex.Lock()
+	defer m.unreadyMapMutex.Unlock()
+	podInfo, found := m.unreadyInterfaces[ovsPort]
+	if !found {
+		klog.InfoS("OVS port is not found", "ovsPort", ovsPort)
+		return nil
+	}
+
+	ofPort, _ := m.ovsPortStatusMap[ovsPort]
+	if err := m.updatePodFlows(ovsPort, int32(ofPort), podInfo); err != nil {
+		klog.ErrorS(err, "Failed to update Pod's OpenFlow entries, requeue the PortStatus message after 10s", "PodName", podInfo.podName, "PodNamespace", podInfo.podNamespace, "OVSPort", ovsPort)
```

Comment:
I didn't find where 10 seconds is defined, maybe we should use constants for easier maintenance?

Reply:
This log is removed in the latest change.